flatfs.go

```diff
@@ -404,6 +589,127 @@ func (fs *Datastore) walkTopLevel(path string, reschan chan query.Result) error
 	return nil
 }
 
+// calculateDiskUsage tries to read the DiskUsageFile for a cached
+// diskUsage value, otherwise walks the datastore files.
+func (fs *Datastore) calculateDiskUsage() error {
+	// Try to obtain a previously stored value from disk
+	if persDu := fs.readDiskUsageFile(); persDu > 0 {
+		atomic.StoreInt64(&fs.diskUsage, persDu)
+		return nil
+	}
+
+	// Otherwise walk the datastore files
+	var du int64
+	err := filepath.Walk(fs.path,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			du += info.Size()
+			return nil
+		})
+
+	if err != nil {
+		log.Errorf("cannot calculate disk usage: %s", err)
+		return err
+	}
+
+	atomic.StoreInt64(&fs.diskUsage, du)
+	return nil
+}
+
+func fileSize(path string) int64 {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return 0
+	}
+	return fi.Size()
+}
+
+// updateDiskUsage reads the size of path and atomically
+// increases or decreases the diskUsage variable.
+// setting add to false will subtract from disk usage.
+func (fs *Datastore) updateDiskUsage(path string, add bool) {
+	fsize := fileSize(path)
+	if !add {
+		fsize = -fsize
+	}
+
+	if fsize != 0 {
+		newDu := atomic.AddInt64(&fs.diskUsage, fsize)
+		fs.checkpointDiskUsage(newDu)
+	}
+}
+
+func (fs *Datastore) checkpointDiskUsage(newDuInt int64) {
```

Comment:
There *is* a serious drawback to this. We could write a bunch of data but crash before checkpointing. The old way was *much* safer because we didn't leave the file on disk while it could be invalid.

Reply:
You would need to crash precisely in the write operation which would do a checkpoint while having written a really large entry (with respects to the current size) for the checkpoint to be really skewed from actual size. I think it's not so much of a big deal as soon as the size of the datastore is sufficiently big to the size of the entries written to it.

Reply:
Fair enough.